File: python/261-graph-valid-tree.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        uf = UnionFind(n)
        for edge in edges:
            if uf.union(edge[0], edge[1]) == False:
                return False
            
        return uf.isTree()

class UnionFind:
    def __init__(self, N) -> None:
        self.parent = [i for i in range(N)]
        self.rank = [1] * N
        
    def find(self, num):
        if self.parent[num] == num:
            return num
        return self.find(self.parent[num])
    
    def union(self,u, v):
        pu, pv = self.find(u), self.find(v)
        if pu == pv:
            return False

        if self.rank[pu] >= self.rank[pv]:
            self.parent[pv] = pu
            self.rank[pu] +=self.rank[pv]
        else:
            self.parent[pu] = pv
            self.rank[pv] +=self.rank[pu]
        return True
        
    
    def isTree(self):
        print(self.parent)
        print(self.rank)
        root = 0
        for i in range(len(self.parent)):
            if self.parent[i] == i:
                root += 1
                if root > 1:
                    return False
                
        return True

File: tao/java/261-graph-valid-tree.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        int[] rank = new int[n];
        int[] parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 1;
        }

        for (int[] edge : edges) {
            int node0 = getParent(edge[0], parent);
            int node1 = getParent(edge[1], parent);

            if (node0 == node1)
                return false;

            if (rank[node0] >= rank[node1]) {
                rank[node0] += rank[node1];
                parent[edge[1]] = node0;
                parent[node1] = node0; // need to update parent node as well
            } else {
                rank[node1] += rank[node0];
                parent[edge[0]] = node1;
                parent[node0] = node1;
            }
        }
        // System.out.println(Arrays.toString(parent));

        int counter = 0;
        for (int i = 0; i < n; i++) {
            if (parent[i] == i) {
                counter++;
                if (counter > 1)
                    return false;
            }
        }

        return true;
    }

    private int getParent(int node, int[] parent) {
        while (node != parent[node]) {
            node = parent[node];
        }
        return node;
    }
}